Previous warnings in chansrv_fuse.c are now errors due to -Werror

There's a difference in latest [version 0.9.14](https://github.com/neutrinolabs/xrdp/releases/tag/v0.9.14) compared to [0.9.13.1](https://github.com/neutrinolabs/xrdp/releases/tag/v0.9.13.1) coming from commit 7e58209 by @aquesnel, which now turns all warnings into errors on Linux (i.e. with `-Werror` compiler option). However, thus previous warnings in chansrv_fuse.c are now turned into errors and thus are blocking the build of the software.

The relevant part of my build log is:

```
Making all in chansrv
make[3]: Entering directory '/usr/src/tmp/xrdp-0.9.14/sesman/chansrv'
i586-slackware-linux-gcc -DHAVE_CONFIG_H -I. -I../..  -DXRDP_CFG_PATH=\"/etc/xrdp\" -DXRDP_SBIN_PATH=\"/usr/sbin\" -DXRDP_SHARE_PATH=\"/usr/share/xrdp\" -DXRDP_PID_PATH=\"/var/run\" -DXRDP_SOCKET_PATH=\"/tmp/.xrdp\" -I../../common  -DXRDP_FUSE -D_FILE_OFFSET_BITS=64 -I/usr/include/fuse  -DXRDP_OPUS -DXRDP_MP3LAME    -O3 -march=i586 -mtune=i686 -Wall -Wwrite-strings -Werror -O2 -c -o chansrv_fuse.o chansrv_fuse.c
chansrv_fuse.c: In function 'xfuse_cb_readdir':
chansrv_fuse.c:1610:20: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
     else if ((dh = (struct xfs_dir_handle *) fi->fh) == NULL)
                    ^
chansrv_fuse.c: In function 'xfuse_cb_read':
chansrv_fuse.c:2098:15: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
     if ((fh = (XFUSE_HANDLE *)fi->fh) == NULL)
               ^
chansrv_fuse.c: In function 'xfuse_cb_write':
chansrv_fuse.c:2171:15: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
     if ((fh = (XFUSE_HANDLE *)fi->fh) == NULL)
               ^
chansrv_fuse.c: In function 'xfuse_cb_releasedir':
chansrv_fuse.c:2439:33: error: cast to pointer from integer of different size [-Werror=int-to-pointer-cast]
     struct xfs_dir_handle *dh = (struct xfs_dir_handle *) fi->fh;
                                 ^
cc1: all warnings being treated as errors
Makefile:535: recipe for target 'chansrv_fuse.o' failed
make[3]: *** [chansrv_fuse.o] Error 1
make[3]: Leaving directory '/usr/src/tmp/xrdp-0.9.14/sesman/chansrv'
Makefile:706: recipe for target 'all-recursive' failed
make[2]: *** [all-recursive] Error 1
make[2]: Leaving directory '/usr/src/tmp/xrdp-0.9.14/sesman'
Makefile:481: recipe for target 'all-recursive' failed
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory '/usr/src/tmp/xrdp-0.9.14'
Makefile:413: recipe for target 'all' failed
make: *** [all] Error 2
```

I'd like to make it clear that I'm trying to build for Slackware 14.2 with GCC 5.5.0, which is quite old. However, building on Slackware-current with GCC 9.3.0 rends even more warnings now turned into errors. It doesn't seem to me that all warnings were fixed, so `-Werror` could be enabled by default on Linux. Should this option remain by default or is there are plan to fix all the warnings?

So @aquesnel, I configure like this:

```
CFLAGS="-O3 -march=i586 -mtune=i686" \
CXXFLAGS="-O3 -march=i586 -mtune=i686" \
./configure --build=i586-slackware-linux \
            --host=i586-slackware-linux \
            --prefix=/usr \
            --libdir=/usr/lib \
            --sysconfdir=/etc \
            --localstatedir=/var \
            --docdir=/usr/doc \
            --mandir=/usr/man \
            --disable-silent-rules \
            --disable-dependency-tracking" \
            --enable-shared=yes \
            --enable-static=no \
            --enable-ipv6 \
            --enable-jpeg \
            --enable-fuse \
            --enable-opus \
            --enable-mp3lame \
            --enable-pixman \
            --enable-painter \
            --enable-rfxcodec
```

The actual build script could be seen here - [xrdp.SLackBuild](https://github.com/gdsotirov/xrdp.SlackBuild/blob/master/xrdp.SlackBuild). I enable the same (i.e. fuse, mp3lame and opus), but I also enable other features (e.g. IPv6, JPEG, pixman, painter and rfxcodec) and set paths, which is necessary for the build of a general purpose system-wide package. The exact differences with your compiler command are the following:

```diff
--- aquesnel
+++ gdsotirov
@@ -1,14 +1,14 @@
-gcc \
+i586-slackware-linux-gcc \
 -DHAVE_CONFIG_H -I. -I../.. \
 -DXRDP_CFG_PATH=\"/etc/xrdp\" \
--DXRDP_SBIN_PATH=\"/usr/local/sbin\" \
--DXRDP_SHARE_PATH=\"/usr/local/share/xrdp\" \
+-DXRDP_SBIN_PATH=\"/usr/sbin\" \
+-DXRDP_SHARE_PATH=\"/usr/share/xrdp\" \
 -DXRDP_PID_PATH=\"/var/run\" \
 -DXRDP_SOCKET_PATH=\"/tmp/.xrdp\" \
 -I../../common \
 -DXRDP_FUSE -D_FILE_OFFSET_BITS=64 \
 -I/usr/include/fuse \
 -DXRDP_OPUS -DXRDP_MP3LAME \
--g -O2 \
--Wall -Wwrite-strings -Werror -MT chansrv_fuse.o -MD -MP -MF .deps/chansrv_fuse.Tpo \
+-O3 -march=i586 -mtune=i686 \
+-Wall -Wwrite-strings -Werror -MT chansrv_fuse.o -MD -MP -MF .deps/chansrv_fuse.Tpo -O2 \
 -c -o chansrv_fuse.o chansrv_fuse.c
```

_Note_: Command `i586-slackware-linux-gcc` on Slackware is just a symbolic link to GCC.

But I do not think these differences are important. And of course I was able to reproduce the same warnings (i.e. errors) with the same compiler command as you. So what's your compiler version? And the next things that come to mind are: What version of FUSE you're using? What OS and version? I'm trying to build package for Slackware 14.2 (released 2016-07-01), so I'm building with old GCC 5.5.0 and old FUSE version - 2.9.5.

However, I tried the same build on Slackware-current with newer GCC 9.3.0 and newer FUSE 2.9.9, but with removed `-Werror` option I have the same warnings plus one more:

```
log.c:558:30: warning: ‘%.2d’ directive output may be truncated writing between 2 and 11 bytes into a region of size between 9 and 16 [-Wformat-truncation=]
chansrv_fuse.c:1610:20: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
chansrv_fuse.c:2098:15: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
chansrv_fuse.c:2171:15: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
chansrv_fuse.c:2439:33: warning: cast to pointer from integer of different size [-Wint-to-pointer-cast]
```

I'm totally for fixing all compilation warnings, which is necessary, before enforcing `-Werror`. However, this is what I get in my environments and searching for the reason I found [your revision](https://github.com/neutrinolabs/xrdp/commit/7e58209b195052205831d53a9f4cd0b4604a293d). I had the same warnings for 0.9.13, but they were just warnings back then. Let me know if you need other information from me. Let's first try to understand why I have warnings and you do not.

--- a/.travis.yml
+++ b/.travis.yml
@@ -82,17 +82,17 @@ max_x86_deps: &max_x86_deps
   - libxrandr-dev:i386
   - libxrender-dev:i386
   - openssl:i386
+  - libfuse-dev:i386
 
 # No --enable-pixman to allow testing the replacement code
-# No --enable-fuse due to failing libfuse-dev:i386 package install
 max_x86_conf: &max_x86_conf
   env:
-    - CONF_FLAGS="--enable-ipv6 --enable-jpeg --disable-fuse --enable-mp3lame
+    - CONF_FLAGS="--enable-ipv6 --enable-jpeg --enable-fuse --enable-mp3lame
                   --enable-fdkaac --enable-opus --enable-rfxcodec --enable-painter
                   --disable-pixman --host=i686-linux"
     - PKG_CONFIG_PATH=/usr/lib/i386-linux-gnu/pkgconfig
-    - CFLAGS=-m32
-    - LDFLAGS=-m32
+    - CFLAGS="-m32"
+    - LDFLAGS="-m32"
   addons:
     apt:
       packages:
@@ -154,5 +154,5 @@ matrix:
 script:
   - ./bootstrap
   - ./configure $CONF_FLAGS
-  - make CFLAGS="$CFLAGS"
+  - make
   - test -z "$DISTCHECK" || make distcheck

--- a/configure.ac
+++ b/configure.ac
@@ -406,6 +406,9 @@ echo "  exec_prefix             $exec_prefix"
 echo "  libdir                  $libdir"
 echo "  bindir                  $bindir"
 echo "  sysconfdir              $sysconfdir"
+echo ""
+echo "  CFLAGS = $CFLAGS"
+echo "  LDFLAGS = $LDFLAGS"
 
 # xrdp_configure_options.h will be written to the build directory, not the source directory
 echo '#define XRDP_CONFIGURE_OPTIONS \' > ./xrdp_configure_options.h

--- a/sesman/chansrv/chansrv_fuse.c
+++ b/sesman/chansrv/chansrv_fuse.c
@@ -1607,7 +1607,7 @@ static void xfuse_cb_readdir(fuse_req_t req, fuse_ino_t ino, size_t size,
         log_error("inode %ld is not valid", ino);
         fuse_reply_err(req, ENOENT);
     }
-    else if ((dh = (struct xfs_dir_handle *) fi->fh) == NULL)
+    else if ((dh = (struct xfs_dir_handle *) (tintptr) fi->fh) == NULL)
     {
         /* something seriously wrong somewhere! */
         fuse_reply_buf(req, 0, 0);
@@ -2095,7 +2095,7 @@ static void xfuse_cb_read(fuse_req_t req, fuse_ino_t ino, size_t size,
 
     log_debug("want_bytes %zd bytes at off %lld", size, (long long) off);
 
-    if ((fh = (XFUSE_HANDLE *)fi->fh) == NULL)
+    if ((fh = (XFUSE_HANDLE *) (tintptr) fi->fh) == NULL)
     {
         fuse_reply_err(req, EINVAL);
     }
@@ -2168,7 +2168,7 @@ static void xfuse_cb_write(fuse_req_t req, fuse_ino_t ino, const char *buf,
     log_debug("write %zd bytes at off %lld to inode=%ld",
               size, (long long) off, ino);
 
-    if ((fh = (XFUSE_HANDLE *)fi->fh) == NULL)
+    if ((fh = (XFUSE_HANDLE *) (tintptr) fi->fh) == NULL)
     {
         log_error("file handle fi->fh is NULL");
         fuse_reply_err(req, EINVAL);
@@ -2436,7 +2436,7 @@ static void xfuse_cb_opendir(fuse_req_t req, fuse_ino_t ino,
 static void xfuse_cb_releasedir(fuse_req_t req, fuse_ino_t ino,
                                 struct fuse_file_info *fi)
 {
-    struct xfs_dir_handle *dh = (struct xfs_dir_handle *) fi->fh;
+    struct xfs_dir_handle *dh = (struct xfs_dir_handle *) (tintptr) fi->fh;
     xfs_closedir(g_xfs, dh);
     fuse_reply_err(req, 0);
 }

